File: Queue/622. 设计循环队列.js
/**
 * @param {number} k
 */
 var MyCircularQueue = function(k) {
    this.list = new Array(k)
    this.size = k
    this.head = -1
    this.tail = -1
};

/** 
 * @param {number} value
 * @return {boolean}
 */
MyCircularQueue.prototype.enQueue = function(value) {
    if(this.isFull()) return false;
    if(this.isEmpty()){
        this.head++;
        this.tail++;
        this.list[this.tail] = value;
        return true;
    }
    this.tail = (this.tail + 1) % this.size
    this.list[this.tail] = value;
    return true;
};

/**
 * @return {boolean}
 */
MyCircularQueue.prototype.deQueue = function() {
    if(this.isEmpty()) return false;
    if(this.head === this.tail){
        this.head = -1;
        this.tail = -1;
        return true
    }
    this.head = (this.head+1) % this.size;
    return true;
};

/**
 * @return {number}
 */
MyCircularQueue.prototype.Front = function() {
    return this.isEmpty() ? -1 : this.list[this.head]
};

/**
 * @return {number}
 */
MyCircularQueue.prototype.Rear = function() {
    return this.isEmpty() ? -1 : this.list[this.tail]
};

/**
 * @return {boolean}
 */
MyCircularQueue.prototype.isEmpty = function() {
    return this.head == -1 && this.tail == -1
};

/**
 * @return {boolean}
 */
MyCircularQueue.prototype.isFull = function() {
    return (this.tail+1) % this.size == this.head
};

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * var obj = new MyCircularQueue(k)
 * var param_1 = obj.enQueue(value)
 * var param_2 = obj.deQueue()
 * var param_3 = obj.Front()
 * var param_4 = obj.Rear()
 * var param_5 = obj.isEmpty()
 * var param_6 = obj.isFull()
 */

File: Python/队列/622. 设计循环队列.py
# -*- codeing = utf-8 -*-
# @Time : 2023-9-6 16:14
# @Author : 刘永奇
# @File : 622. 设计循环队列.py
# @Software : PyCharm
class MyCircularQueue:

    def __init__(self, k: int):
        self.font = 0
        self.cnt = 0
        self.tail = 0
        self.k = k
        self.queue = [None]*k


    def enQueue(self, value: int) -> bool:
        if self.cnt == self.k:return False
        self.queue[self.tail] = value
        self.tail += 1
        self.cnt += 1
        if self.tail == self.k:
            self.tail = 0
        return True

    def deQueue(self) -> bool:
        if self.cnt == 0: return False
        self.font += 1
        self.cnt -= 1
        if self.font == self.k:
            self.font = 0
        return True

    def Front(self) -> int:
        if self.cnt == 0:return -1
        return self.queue[self.font]


    def Rear(self) -> int:
        if self.cnt == 0:return -1
        if self.tail == 0:
            return self.queue[self.k-1]
        return self.queue[self.tail - 1]


    def isEmpty(self) -> bool:
        return self.cnt == 0


    def isFull(self) -> bool:
        return self.cnt == self.k



# Your MyCircularQueue object will be instantiated and called as such:
# obj = MyCircularQueue(k)
# param_1 = obj.enQueue(value)
# param_2 = obj.deQueue()
# param_3 = obj.Front()
# param_4 = obj.Rear()
# param_5 = obj.isEmpty()
# param_6 = obj.isFull()